[Doran] Weather step3

hwj0623 opened the pull request:
### 날씨앱 3단계 구현했습니다.
 - 상세내역은 readme에 기술하였습니다.
 - http 통신에 대한 ATS 정책 설정중에 다른 설정 방식이 잘 작동되지 않아서 `Allow Arbitrary Loads` 로 설정하였습니다.

hwj0623 committed:
TableCell identifier를 subtitleCell로 변경

hwj0623 committed:
커스템 셀 생성 및 오토레이아웃 설정

hwj0623 committed:
새로 생성한 커스템 셀이 테이블에 나오도록 변경

hwj0623 committed:
date/subtitle label 색상 변경

hwj0623 committed:
배경 이미지에 대한 ContentMode를 AspectToFill로 변경

hwj0623 committed:
enum에 image 추가

hwj0623 committed:
App Transport Policy 위반사항인 http 통신을 위한 설정 추가

hwj0623 committed:
URL 을 통해 weather 정보 네트워크로 받는 기능 추가

hwj0623 committed:
Image 정보에 따라 table Cell의 배경 이미지 적용하는 기능 추가
- image 정보 key값이 없는 이미지 렌더링에 대한 버그 수정
- 하드코딩 개선

hwj0623 committed:
reusable cell에 대한 기존 셀 id와 변경된 셀 id에 대한 하드코딩 제거

hwj0623 committed:
fix : 공휴일정보 unwrapping시 var -> let 으로 변경

hwj0623 committed:
HolidayTableViewCell에 prepareForReuse 메서드 오버라이딩

hwj0623 committed:
Readme에 학습 정리 추가

hwj0623 committed:
http ATS 정책 설정 중복 제거

hwj0623 committed:
http통신에 대한 개발단계 ATS 정책 설정내용 추가

godrm commented on WeatherApp/WeatherApp/HolidayViewController.swift:
```diff
@@ -31,18 +31,28 @@ extension HolidayViewController: UITableViewDataSource {
+            }else {
+                cell.backgroundColor = .gray
+            }
         }
```
cell 처리하는게 이렇게 길어지면 DataSource에 있을께 아니라 클로저도 cell 내부에 있는게 좋습니다. 
자기 자신의 속성을 채우는 일이니까요.

hwj0623 replied:
뷰 컨트롤러에 연계된 메서드만 없다면 그렇게 하는게 더 간결해지겠군요. 수정하겠습니다.

godrm replied:
오히려 연계된 메서드가 있으면 이상한거죠.

godrm commented on WeatherApp/WeatherApp/HolidayList.swift:
```diff
@@ -33,15 +37,28 @@ struct HolidayList {
     }
+    
+    private func checkNilForImage(_ holiday: [String: Any]) -> Bool {
+        guard holiday[KeyInfo.image.rawValue] != nil else {
```
`return holiday[KeyInfo.image.rawValue] == nil` 형태를 먼저 생각해보세요.

godrm commented on WeatherApp/WeatherApp/HolidayList.swift:
```diff
@@ -33,15 +37,28 @@ struct HolidayList {
         for holiday in holidays {
-            date = holiday[KeyInfo.date.rawValue] as! String
-            subtitle = holiday[KeyInfo.subtitle.rawValue] as! String
+            var dictionaryElement = [String: String]()
```
dictionaryElement를 띄어서 생각할 수 있었으면 이것도 타입으로 만드는게 어떨까 생각해보세요.
Dictionary는 일반적인 타입이고, 원하는 데이터만 있는지 없는지 판단하려면 타입을 만드는게 좋습니다.

hwj0623 replied:
원래 별도의 구조체 `Holiday`를 만들어서 `[Holiday]` 의 형태로 가져갈까 했는데, 문제 요구사항이 Array<Dictionary<String:String>> 형태의 자료를 요구하는 것 같아서 놔뒀었습니다.
문제 요구사항을 조금 변경해도 상관없는지요?

godrm replied:
JSON에서 1차적으로 바뀌는 구조가 정해져 있던 것 뿐입니다. 
문제 상황보다 더 좋은 구조로 개선하는 것은 어떤 스텝에서든지 중요한 포인트입니다.

godrm commented on WeatherApp/HolidayJsonDTO.swift:
```diff
@@ -11,24 +11,17 @@ import Foundation
+                return
+            }
+            self.data = jsonData
+        }catch {
```
catch 이후에는 반드시 흐름 제어가 있어야 합니다.
이 객체를 생성하지 못하면 어떻게 해야할까요?
어떤 영향을 줄 수 있을까요? 객체를 실패했다는 것을 알 수 있는게 좋습니다.

godrm commented on WeatherApp/HolidayJsonDTO.swift:
```diff
@@ -11,24 +11,17 @@ import Foundation
     }
+    
     var initiateHolidayData: Data {
         return self.data
```
이렇게 생성자에서 만들고 값을 가져갈꺼라면 그냥 static 메소드로 만드는 건 어떤가요? 꼭 인스턴스로 만들어야 하나요?

hwj0623 committed:
네트워크 데이터 로드에 대한 예외처리 추가

hwj0623 committed:
fix : AppDelegate 내 모델 주입 시점 변경

hwj0623 committed:
예외처리에 대한 메시지 enum 추가

hwj0623 committed:
fix : image Nil check 개선, 이니셜라이저 수정

hwj0623 committed:
fix : 테이블 콘텐츠 데이터 바인딩 클로저 포맷을 분리하여 Cell 클래스로 이동

hwj0623 committed:
fix : 예외처리에 대한 옵저버 등록, 핸들러 추가, 테이블 뷰 예외처리 로직 추가

hwj0623 committed:
개선사항 관련 학습 내용 추가

hwj0623 committed:
dictionaryElement를 별도 구조체-Holiday-로 분리

hwj0623 committed:
개선된 결과화면 추가

hwj0623 committed:
fix : 중복 메서드 삭제

hwj0623 committed:
fix : 하드코딩 내용 개선

hwj0623 committed:
fix : tab 간격 일부 수정

hwj0623 committed:
fix : 테이블 뷰 화면 하단 여백 수정, 이미지 재 업로드

hwj0623 committed:
fix : 모델 주입/생성은 값이 없을때만 하도록 처리

hwj0623 commented:
피드백 반영하여 개선했습니다.
- AlertController를 활용한 예외처리 로직을 위해 AppDelegate의 application 메서드에서의 setter 주입이 아니라 viewDidAppear에서 모델의 데이터를 받아오도록 하였습니다.
- 이후에 백그라운드로 간 앱을 다시 활성화 할 때 `applicationWillEnterForeground`에서 주입 시도하도록 하였습니다.

godrm commented on WeatherApp/WeatherApp/HolidayList.swift:
```diff
@@ -10,55 +10,52 @@ import Foundation
+        return format(holidays.count)
     }
     
     func receiveTableViewContentFormat( format: (_ date: String, _ subtitle: String, _ image: String?) -> Void, rowAt: Int) {
```
Holiday 타입을 만들었으면 클로저를 활용할 때도 Holiday를 직접 활용하는게 좋습니다. 참고하세요.

godrm commented on WeatherApp/WeatherApp/HolidayTableViewCell.swift:
```diff
@@ -19,4 +19,23 @@ class HolidayTableViewCell: UITableViewCell {
+                self.backgroundColor = .gray
+            }
+        }
+        return tableViewContentFormat
```
이렇게 computed property로 구현하면 호출할 때마다 매번 클로저가 만들어집니다. 
함수로 만들고 함수 자체를 클로저 대신에 넘겨도 되지 않을까요?

godrm commented on WeatherApp/HolidayJsonDTO.swift:
```diff
@@ -10,19 +10,32 @@ import Foundation
+        if flag {
+            holidayURL = URL.init(string: URLInfo.correctURL.rawValue)!
+        }else {
+            holidayURL = URL.init(string: URLInfo.wrongURL.rawValue)!
```
개인적으로는 enum : String 으로 선언해서 rawValue까지 쓰면 너무 길어지는 표현같습니다. 
간편하려고 쓰는건데 더 길어지는 거면, 꼭 enum 으로 선언할 필요는 없지 않을까요?

godrm commented on WeatherApp/WeatherApp/HolidayViewController.swift:
```diff
@@ -13,11 +13,60 @@ class HolidayViewController: UIViewController {
+            if retry.isEnabled {
+                self.retryNetworkReceive()
+                DispatchQueue.main.async {
+                    self.holidayTableView.reloadData()
```
리로드 하는 시점은 여기보다는 네트워크에서 응답이 왔을때가 적합한 것 같습니다.
항상 뷰를 업데이트하는 것은 이벤트를 처리하는 곳에서 하지마세요. 
이벤트를 처리하는 곳은 모델을 바꾸는 요청만 하고, 모델이 바뀔때 그리는 흐름을 이어가세요.

hwj0623 committed:
fix : 테이블 뷰 컨텐츠 관련된 클로저를 함수로 대체

hwj0623 committed:
fix : 일부 rawValue를 StringInterpolation으로 변경

hwj0623 committed:
fix : 테이블 뷰의 reloadData를 model 변경시점에 NotificationCenter를 통해서 이뤄지도록 개선
- NotificationName 대문자 시작으로 변경
- (아카이빙 로직이 없으므로) background에서 복귀시 올바른 주소에 접근하여 모델 초기화 하도록 수정
- "result"를 enum 값의 StringInterpolation으로 변경

godrm commented on WeatherApp/WeatherApp/WeatherEnums.swift:
```diff
@@ -37,8 +37,11 @@ enum ButtonMessage: String, CustomStringConvertible {
-    case wrongResult = "result"
+    case result = "result"
+    var description: String {
+        return self.rawValue
```
어차피 문자열값이고, 문자열로 쓸꺼니까 URLInfo가 enum 타입이 아니라 struct 이고 static let 정도만 되고 충분할 것 같은데
description 을 구현하고 다시 StringInterpolation을 구현한 건 약간 오버같다는 의미였습니다. 참고하세요.

hwj0623 replied:
이전단계에서는 비슷한 경우에 struct로 했던것 같은데, 피드백 의도를 잘못 파악했네요. 마지막 내용만 수정해서 재차 PR 드리겠습니다.

hwj0623 commented:
피드백 사항 재차 반영했습니다.
- 이벤트 핸들러에서 UI변경 처리 -> notificationCenter를 거치도록 수정하였습니다.
- enum String의 rawValue는 CustomStringConvertible 채택 후에 StringInterpolation으로 표현 변경하였습니다.
- 클로저는 함수로 변경했습니다.
